File: src/test/java/io/javalin/examples/AsyncExample.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.examples;

import io.javalin.Javalin;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.CompletableFuture;
import javax.servlet.AsyncContext;

public class AsyncExample {

    public static void main(String[] args) {

        Javalin app = Javalin.create().port(5454).start();

        app.get("/test-custom", ctx -> {
            AsyncContext asyncContext = ctx.request().startAsync();
            simulateAsyncTask(() -> {
                ctx.status(418);
                asyncContext.complete();
            });
        });

        app.get("/test-async", ctx -> ctx.async(() -> {
            CompletableFuture<Void> future = new CompletableFuture<>();
            simulateAsyncTask(() -> {
                ctx.status(418);
                future.complete(null);
            });
            return future;
        }));

    }

    private static void simulateAsyncTask(Runnable runnable) {
        new Timer().schedule(
            new TimerTask() {
                public void run() {
                    runnable.run();
                }
            },
            1000
        );
    }

}

File: src/test/kotlin/io/javalin/examples/AsyncExample.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.examples

import io.javalin.Javalin
import java.util.*
import java.util.concurrent.CompletableFuture


fun main(args: Array<String>) {

    val app = Javalin.create().port(5454).start()

    app.get("/test-custom") { ctx ->
        val asyncContext = ctx.request().startAsync()
        simulateAsyncTask({
            ctx.status(418)
            asyncContext.complete()
        })
    }

    app.get("/test-async") { ctx ->
        ctx.async {
            val future = CompletableFuture<Void>()
            simulateAsyncTask({
                ctx.status(418)
                future.complete(null)
            })
            future
        }
    }

}

private fun simulateAsyncTask(runnable: () -> Unit) {
    Timer().schedule(
            object : TimerTask() {
                override fun run() {
                    runnable.invoke()
                }
            },
            1000
    )
}
